hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/impl/ITestS3APutIfMatch.java

```diff
@@ -0,0 +1,142 @@
+        Configuration conf = super.createConfiguration();
+        S3ATestUtils.disableFilesystemCaching(conf);
+        removeBaseAndBucketOverrides(conf,
+            MULTIPART_SIZE,
+            UPLOAD_PART_COUNT_LIMIT);
+        conf.setLong(MULTIPART_SIZE, MPU_SIZE);
+        conf.setLong(UPLOAD_PART_COUNT_LIMIT, 2);
+        conf.setLong(MIN_MULTIPART_THRESHOLD, MULTIPART_MIN_SIZE);
+        conf.setInt(MULTIPART_SIZE, MULTIPART_MIN_SIZE);
+        conf.set(FAST_UPLOAD_BUFFER, getBlockOutputBufferName());
+        return conf;
+    }
+
+    @Override
+    public void setup() throws Exception {
+        super.setup();
+        conf = createConfiguration();
+        skipIfNotEnabled(conf, FS_S3A_CONDITIONAL_FILE_CREATE,
+                "Skipping IfNoneMatch tests");
+    }
+
+    protected String getBlockOutputBufferName() {
+        return FAST_UPLOAD_BUFFER_ARRAY;
+    }
+
+    /**
+     * Create a file using the PutIfMatch feature from S3
+     * @param fs filesystem
+     * @param path       path to write
+     * @param data source dataset. Can be null
+     * @throws IOException on any problem
+     */
+    private static void createFileWithIfNoneMatchFlag(FileSystem fs,
+                                                      Path path,
+                                                      byte[] data,
+                                                      String ifMatchTag) throws Exception {
+          FSDataOutputStreamBuilder builder = fs.createFile(path);
+          builder.must(FS_S3A_CONDITIONAL_FILE_CREATE, ifMatchTag);
+          FSDataOutputStream stream = builder.create().build();
+          if (data != null && data.length > 0) {
+              stream.write(data);
+          }
+          stream.close();
+          IOUtils.closeStream(stream);
+    }
+
+    @Test
+    public void testPutIfAbsentConflict() throws IOException {
+        FileSystem fs = getFileSystem();
+        Path testFile = methodPath();
+
+        fs.mkdirs(testFile.getParent());
+        byte[] fileBytes = dataset(TEST_FILE_LEN, 0, 255);
+
+        try {
+          createFileWithIfNoneMatchFlag(fs, testFile, fileBytes, "*");
+          createFileWithIfNoneMatchFlag(fs, testFile, fileBytes, "*");
+        } catch (Exception e) {
+          Assert.assertEquals(RemoteFileChangedException.class, e.getClass());
+
+          S3Exception s3Exception = (S3Exception) e.getCause();
+          Assertions.assertThat(s3Exception.statusCode()).isEqualTo(412);
+        }
+    }
+
+
+    @Test
+    public void testPutIfAbsentLargeFileConflict() throws IOException {
+        FileSystem fs = getFileSystem();
+        Path testFile = methodPath();
+
+        // enough bytes for Multipart Upload
+        byte[] fileBytes = dataset(6 * _1MB, 'a', 'z' - 'a');
+
+        try {
+            createFileWithIfNoneMatchFlag(fs, testFile, fileBytes, "*");
+            createFileWithIfNoneMatchFlag(fs, testFile, fileBytes, "*");
+        } catch (Exception e) {
```

Comment:
this must use intercept() again. Not just for its compactness but it will raise a exception if the lambda exception *doesn't* fail.

Reply:
updated to 
```
        RemoteFileChangedException firstException = intercept(RemoteFileChangedException.class,
                () -> createFileWithIfNoneMatchFlag(fs, testFile, fileBytes, "*"));
        assertS3ExceptionStatusCode(412, firstException);

        RemoteFileChangedException secondException = intercept(RemoteFileChangedException.class,
                () -> createFileWithIfNoneMatchFlag(fs, testFile, fileBytes, "*"));
        assertS3ExceptionStatusCode(412, secondException);
```